orttraining/orttraining/core/optimizer/graph_transformer_utils.cc

```diff
@@ -94,7 +94,11 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
       ORT_THROW_IF_ERROR(rule_transformer->Register(std::make_unique<InsertSoftmaxCrossEntropyLossOutput>()));
 
       // Remove duplicate nodes. Must be applied before any recompute transformations.
-      transformers.emplace_back(std::make_unique<CommonSubexpressionEliminationApplyOnce>(compatible_eps));
+      if (config.gelu_recompute || config.attn_dropout_recompute || config.transformer_layer_recompute) {
```

Comment:
As the comment (legacy code) said, any legacy recompute transformations MUST be after CSE, otherwise those recomputed nodes will be optimized out by the CSE again, so I guess at that time, we add a new CSE called CSEApplyOnce. 

While if we don't enable any of the legacy recompute transformations, we don't need have this restriction. The reason I loose the restriction for legacy-recompute-not-enabled path is I found there are still opportunities to do CSE a second or third times for the model I was running (Roberta) but we did not do that, then some subsequent optimization (for example the ComputeOptimizer is not applicable.

Reply:
Related PR https://github.com/microsoft/onnxruntime/pull/5504